### ISSUE TITLE
Provide ability to open dynamically created components

### ISSUE THREAD

**Issue by herwix** [shlomiassaf/angular2-modal#200]
IMPORTANT: Please provide a sample using:  

  * V1: http://plnkr.co/edit/hwbmF9?p=preview 
  * V2: http://plnkr.co/edit/7c0pGE?p=preview

* **I'm submitting a ...**
[ ] bug report
[X ] feature request
[ ] question about the decisions made in the repository

* **Do you want to request a *feature* or report a *bug*?**

I need to create a modal component from a module which I have dynamically compiled in the app. It would be awesome if I could simply provide the proper injector that will be used for custom modal creation.

* **What is the current behavior?**

I have look trough the source code and I have not found an easy way to do this. I could probably subclass the overlayRenderer to achieve what I want to do, but there is not really good documentation on how to do that in a way that would be aligned with the intentions of the module (e.g., for compatibility with future versions, etc.)

* **If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem** via
https://plnkr.co or similar (you can use one of the templates above as a starting point).

* **What is the expected behavior?**

overlayConfigFactory optionally also accepts an injector that will be used to create the custom component (not the overlay itself).

* **What is the motivation / use case for changing the behavior?**

Dynamically compiled modules with modal components.

* **Please tell us about your environment:**

- Angular version: 2.0.0
- Browser: [all | Chrome XX | Firefox XX | IE XX | Safari XX | Mobile Chrome XX | Android X.X Web Browser | iOS XX Safari | iOS XX UIWebView | iOS XX WKWebView ]

* **Other information** (e.g. detailed explanation, stacktraces, related issues, suggestions how to fix, links for us to have context, eg. stackoverflow, gitter, etc)

**Comment by JulioC** [comment]
I'm also having issues when trying to open component from dynamically compiled ngModules.

I haven't investigated it yet, but I believe the service on angular2-modal that creates the Component is hierarchically above the loaded module and therefore can't find their component factories.

Both solutions from @herwix looks fine to me

**Comment by shlomiassaf** [comment]
I'm not sure what you mean...

the `open` method of the `Modal` service accepts either a Component, TemplateRef or a string.
You can also attach bindings in the `overlayConfig`.

**Comment by herwix** [comment]
I think that JulioC has the right idea. Basically when you dynamically compile a module than you are possibly loading new components that have not been registered with the hierarchical injector in the main app component (where the defaultOverlay is located). So when you try to open that component it cannot be found by the modal service because the hierarchical injector does not know about the components in the dynamic compiled module. At leasts that's my thinking...

I have been trying to work around that by using the injector of the application as a parent injector for the compilation of my dynamically compile module and using the overlayTarget directive in a dynamically compiled module so that I can reference the overlayTarget by name and use the appropriate injector for the creation of the modal content. But I have not gotten it running as of yet.

I hope the use case is a little bit clearer now. If you have any idea how to make it work that would be greatly appreciated :) And thanks for the effort with the module, it's really helpful and I am using it a lot :)

**Comment by shlomiassaf** [comment]
@herwix You can supply bindings in the `OverlayConfig`, i'm not sure it will help since I don't really know what is the token for a component, maybe it's just the component class maybe is a multi provider with a list of components... dont know.

Can you try setting the `defaultViewContainer` via template and not programatically.
```
<span defaultOverlayTarget></span>
```
See [this example](https://github.com/shlomiassaf/angular2-modal/blob/master/src/demo/app/app.html#L1)

It's kinda like the `overlayTarget` attempt you try to run but with less places to make mistakes.

The template/component hosting the 
```
<span defaultOverlayTarget></span>
```
Must have an injector that already has the dynamic components registered.

**Comment by herwix** [comment]
@shlomiassaf I have used the defaultOverlayTarget in my root app component and that works fine. I still need the modals to work there. So, I wanted to use the overlayTarget to create additional modal outlets for my dynamic components (e.g., register them in the vcRefStore) in a place where I have the right injector. I don't think the directive is being picked up though at the moment. Kinda difficult to debug when it's not your own component though :D. I am close to managing the vcRefStore myself with a copy of the targetOverlay directive to see how things work then... 

To make it work I also need dynamic bindings of names in the vcRefStore so I will probably have to do that anyway. 

Everything would be a lot easier if we could simply provide the right injector though and don't have to make the workaround via viewContainerRefs ;) Then we could still keep using the defaultOverlayTarget but also create components that come from a dynamically compiled module. 

Following would be the method that allows you to create components via viewContainerRef with a custom injector.

`
abstract createComponent<C>(componentFactory: ComponentFactory<C>, index?: number, injector?: Injector, projectableNodes?: any[][]): ComponentRef<C>;`

**Comment by herwix** [comment]
So I have been able to get the right viewContainerRef (with proper injector set) and set it as the viewContainer on the overlayConfig. Now no errors are thrown about not recognizing the component, but the component is still not rendered. 

I don't really see why that is happening, though. Any ideas of what might go wrong?

**Comment by JulioC** [comment]
As I understand, setting `viewContainer` will make the overlay appear inside the given element.

If this interpretation is correct, that is not the desired behavior for my case.

I'll try to provide a plunkr illustrating the problem.

**Comment by shlomiassaf** [comment]
@herwix The [DOMOverlayRenderer](https://github.com/shlomiassaf/angular2-modal/blob/master/src/components/angular2-modal/providers/dom-modal-renderer.ts#L15) is using `ComponentFactoryResolver`, I believe it's injected before your dynamic components are registered. 

Now, `ComponentFactoryResolver. resolveComponentFactory` is not getting any `injector` or `ViewContainerRef` so my guess is that it's still using the old Injector, but I might be wrong... probably wrong :)

**Comment by herwix** [comment]
@shlomiassaf so I finally got it working... the problem was that somehow changeDetection wasn't running on the container after it was created... strangely only with components from a different viewContainerRef. Might not be the complete solution but for me it seems to work right now.

My 'fix' is in my [fork](https://github.com/herwix/angular2-modal/blob/master/src/components/angular2-modal/providers/modal.ts#L95). I don't know if it's worthwhile to make a pull request for that. But would be awesome if that could be shipped in the next release.

To sum up my approach:

get a viewContainerRef from a dynamically compiled module and set that manually on the overlayConfig.viewContainer. Then everything should work as expected.

**Comment by shlomiassaf** [comment]
@herwix

Calling `dialogRef.overlayRef.instance.addComponent<T>(ContainerComponent, b, nodes);` includes a call to `changeDetectorRef.detectChanges();` [source code](https://github.com/herwix/angular2-modal/blob/master/src/components/angular2-modal/components/base-dynamic-component.ts#L110)

So, `detectChanges` won't cut it, you must also call `markForCheck`.
So this is where it get's tricky, it can be many things since we are playing with the view tree.

The 1st call is to `detectChanges`, this is a synced call, when it's done we mark to change for changes and this will happen on the next vm turn.

I would like to know the reason, so I know if to add it in the `addComponent` function or just leave it in the modal.

Can you please put up a plunker the illustrates the error, something simple with dynamic compilation. I will try to debug the root cause, should give a good insight on CD as well.

Thanks.

**Comment by herwix** [comment]
@shlomiassaf I am sorry but I am out until late next week. And I don't know if I will have any more time to spend on this, already spent the whole day... maybe @JulioC can set up a case to demonstrate his use case and we can test if the fix is working there as well?

**Comment by JulioC** [comment]
I've made [this plunkr](http://plnkr.co/edit/awmxejiI1cyIOvKX9hgX) illustrating the issue.

`AppComponent` is the defaultOverlayTarget, its module imports both `ModalModule.forRoot()` and `BootstrapModalModule`. It will compile `OtherModule` and create `OtherComponent` in the view.

`OtherComponent` uses `Modal` service to open a `OtherModalComponent`, but fails with:

> Error in ./ModalOverlay class ModalOverlay - inline template:2:14 caused by: No component factory found for OtherModalComponent

I tried following the calls inside Modal.open, but I was unable to identify where it creates the content component.

If there is any mistake on my code, please point it out.

**Comment by herwix** [comment]
Thanks for the plunker! Just to demonstrate my workaround: [plunker](http://plnkr.co/edit/n9kA1MgB90FZSqkpQB9J?p=preview) 

The modal will still be inserted in the same place as before but receive the right injector via the new viewContainerRef. It's not optimal because you have to have a target within the target module but for me that's possible, so I could live with that for the moment. Interestingly, the cdr.markForCheck() does not seem to be needed to be called here. For me it would simply render the modal frame but not the content because the swapComp directive was not receiving its input properties.

**Comment by JulioC** [comment]
I tried your workaround on my application code and somehow it doesn't work.

This part of my application is still really ugly after migrating to ngModules, so I'll rewrite it on the next days. Maybe I can find out why it's not working.

**Comment by herwix** [comment]
If the moda shows up but there is no content (and no error) it might be the same refresh problem that I faced. I also wrote a custom overlaytarget directive which also implements OnChanges but that might not be necessary.

**Comment by JulioC** [comment]
It throws the same exception as before, so I don't think it's the same case as yours.

**Comment by shlomiassaf** [comment]
@herwix @JulioC Thanks for the plunkers.
I'll review it and see what's the best approach to introduce dynamically compiled components.

**Comment by JulioC** [comment]
I refactored the way I lazy load my modules and now the workaround is doing it for me. Sadly I wasn't able to identify what changed fixed it.

I'll keep watching for you solution, but it's not that urgent anymore.

Thanks @shlomiassaf @herwix

**Comment by herwix** [comment]
Just to note for my implementation the call to mark for check would still need to be 'fixed' otherwise my dynamic component is not swapped in. I don't know why though. In the plunker it seems to work fine... but I somehow doubt it is really related to my code because the call in question is before my component gets rendered. (Maybe I am using it in an OnPush part of the component tree and that causes problems?)

Cheers, alex

**Comment by charsleysa** [comment]
@JulioC @herwix I had a similar problem.

I tried to open a modal in a lazy loaded module and I was getting the `No component factory found for xxxxxx` error. I couldn't for life of me figure out why.

My workaround was to setup the main app component like this:
```ts
export class App {
    constructor(
        private overlay: Overlay,
        private viewContainerRef: ViewContainerRef
    ) {
        overlay.defaultViewContainer = viewContainerRef;
    }
}
```

Then, inside my component which resides in the lazy loaded module, I did this (some app specific code omitted):
```ts
export class ContentPage {
    constructor(
        private modal: Modal,
        private router: Router,
        private route: ActivatedRoute,
        private viewContainerRef: ViewContainerRef
    ) { }

    public changeLocation() {
        this.modal.open(ChangeLocationDialog, overlayConfigFactory({}, VEXModalContext, { viewContainer: this.viewContainerRef }));
    }
}
```

This works and places the modal directly inside the body tag and has no rendering issues for me.

Hope this helps!

**Comment by shlomiassaf** [comment]
@JulioC @herwix Didn't forget about this, had to focus on AoT support so project with AoT (all ng-cli's) will be able to use the library. Major change to build process.

I will get to this :)

**Comment by shlomiassaf** [comment]
@JulioC @herwix Here's my first step towards support, allow supplying `injector` to `OverlayConfig`.

The `injector` is optional, if not supplied the injector is taken from the `ViewContainerRef`.

Something like this:
```
          return this.compiler.compileModuleAsync(RuntimeCompiledModule)
            .then(moduleFactory => moduleFactory.create(this.injector))
            .then(module => {
              return this.modal.open(RuntimeCompiledComponent, overlayConfigFactory({}, BSModalContext, {
                injector: module.injector
              }));
            });
```

The overlay is created with the supplied injector which means it's inner `ViewContainerRef` will inherit from that injector. Since all child components inside an overlay are using that `ViewContainerRef` it will work without issues.

One edge case that came in mind is when `ViewContainerRef` is set explicitly in `OverlayConfig`, if no `injector` is set and the `Component` to render is lazy/runtime compiled then it will throw.

So, if you set a view container that was created before a component was registered you also need to supply an injector.

A solution to workaround that, as well as @charsleysa issue, is to re-set the `Overlay. defaultViewContainer` to a new value once a lazy loaded module was loaded.

This is all new so maybe other things might be of help, maybe having a default injector and being able to change that... this is tricky tough it might mess DI structure so I would avoid that for now.

**Comment by shlomiassaf** [comment]
@JulioC regarding you note about supplying a `ViewContainerRef` (string or actual).
If supplied within a call to `modal.open()` it **will not** default to display blocked inside the view container boundaries. i.e: It will not be inside the element.

To display inside an element you must call `inElement(true)`

**Comment by charsleysa** [comment]
@shlomiassaf thanks for your work!

I think in my situation with lazy loaded components passing the injected ViewContainerRef of the component that is instantiating the modal is the best way to do it following your comments as I don't have an instance of the Injector.

**Comment by shlomiassaf** [comment]
@herwix checkout the [demo site](http://shlomiassaf.github.io/angular2-modal/#/bootstrap-demo) in with the bootstrap plugin.

There is a JIT button, it will also allow you to see insertion of JIT compiled component into a previously created view view container ref.

Still not in master, it's in the [dynamic-compilation](https://github.com/shlomiassaf/angular2-modal/tree/dynamic-compilation) branch.

**Comment by shlomiassaf** [comment]
@herwix @JulioC @charsleysa Did you test it, can we close this?

**Comment by JulioC** [comment]
The injector config did it for me. Thanks!

**Comment by charsleysa** [comment]
@shlomiassaf I'm unsure as to the difference between `injector` and `viewContainer` as they both seem to be doing the same thing.

But to answer you question, yes it seems to be working, although I'm just using the injector from the viewContainer:
```ts
export class ContentPage {
    constructor(
        private modal: Modal,
        private router: Router,
        private route: ActivatedRoute,
        private viewContainerRef: ViewContainerRef
    ) { }

    public changeLocation() {
        this.modal.open(ChangeLocationDialog, overlayConfigFactory({}, VEXModalContext, { injector: this.viewContainerRef.injector }));
    }
}
```

**Comment by herwix** [comment]
@shlomiassaf I am currently on vacation, so I am sorry but I won't have time to check until late next week. But as I remember the error that I was describing was actually not really related to the injector but the change detection. So it would be nice to keep this open until I can recheck :)

**Comment by JulioC** [comment]
@charsleysa you can inject `Injector` directly, which is the same as the one from `ViewContainerRef` I believe.

**Comment by vijender1256** [comment]
@herwix @charsleysa Did anyone fix this issue , I am having the same issue :

![image](https://cloud.githubusercontent.com/assets/12258728/20773566/6d6cbcc2-b720-11e6-836a-2288868bf4a7.png)

ConclusionCodeWindow is my component.

Here is my code for Module:

```
import { NgModule }       from "@angular/core";
import { CommonModule }   from "@angular/common";
import { CashierRiskProfileComponent }    from "./cashierriskprofile.component";
import { CashierRiskProfileService }   from "./cashierriskprofile.service";
import { CashierRiskProfileRouting } from "./cashierriskprofile.routing";
import { SharedModule } from "../shared/shared.module";
import { ModalModule } from 'angular2-modal';
import { BootstrapModalModule } from 'angular2-modal/plugins/bootstrap';
import { ConclusionCodeWindow } from './conclusioncode.modal';

@NgModule({
    imports: [
        CommonModule,
        SharedModule,
        CashierRiskProfileRouting,
        ModalModule.forRoot(),
        BootstrapModalModule
    ],
    declarations: [
        CashierRiskProfileComponent,ConclusionCodeWindow
    ],
    providers: [
         CashierRiskProfileService
    ]
})
export class CashierRiskProfileModule { }
```

and for Component here is my code:

```
import { Component, ViewContainerRef, OnInit, Output, EventEmitter, ViewChild } from "@angular/core";
import { Router, ActivatedRoute } from "@angular/router";
import { Modal } from 'angular2-modal/plugins/bootstrap';
import { ConclusionCodeWindow, ConclusionCodeWindowData } from './conclusioncode.modal';
import { Overlay } from 'angular2-modal';

   constructor(overlay: Overlay, vcRef: ViewContainerRef, public modal: Modal) { overlay.defaultViewContainer = vcRef; }

  onStatusChange(status) {
        this.modal.open(ConclusionCodeWindow, new ConclusionCodeWindowData(status));
    }
```

I am getting error No component factory found
